### PR TITLE
Remove throw/catch pattern for Keyword.delete/2 and friends to prevent stacktrace corruption.

### PR DESCRIPTION
**Pull request by potatosalad**
Calling [`Keyword.put/3`](https://github.com/elixir-lang/elixir/blob/e9dfa50c74488000d2c9de71e926cdd78609b3ac/lib/elixir/lib/keyword.ex#L553-L556), [`Keyword.delete/2`](https://github.com/elixir-lang/elixir/blob/e9dfa50c74488000d2c9de71e926cdd78609b3ac/lib/elixir/lib/keyword.ex#L497-L503), and other related functions changes the stacktrace returned from [`:erlang.get_stacktrace/0`](http://erlang.org/doc/man/erlang.html#get_stacktrace-0).

This is due to the throw/catch pattern used in the implementation of these functions.

This pull request helps prevent the following scenario and associated developer confusion:

```elixir
options = []
{exception, options} =
  try do
    :lists.last([])
  catch
    class, reason ->
      options = Keyword.put(options, :reason, reason)
      exception = Exception.normalize(class, reason)
      {exception, options}
  end
exception == %FunctionClauseError{
  args: nil,
  arity: 3,
  clauses: nil,
  function: :delete_key,
  kind: nil,
  module: Keyword
}
```

Notice that the `module`, `function`, and `arity` for the `FunctionClauseError` is `Keyword.delete_key/3` instead of `:lists.last/1`.

In "property test form", the problem could be represented with the following example:

```elixir
try do
  :lists.last([])
rescue _ ->
  frame1 = hd(System.stacktrace())
  _ = Keyword.delete([], :key)
  frame2 = hd(System.stacktrace())
  frame1 == frame2 # currently, this is false, but this pull request changes that
end
```

#### Why does this matter?

Good question...?  It might not.

I ran into this problem initially while using [`sentry-elixir`](https://github.com/getsentry/sentry-elixir) and was confused to see messages logged as `(FunctionClauseError) no function clause matching in Keyword.delete_key/3` instead of `(FunctionClauseError) no function clause matching in MyModule.real_function/1`.

The [fix](https://github.com/potatosalad/sentry-elixir/commit/2ee4c57d7085b5c211cdaff0911cac3720b5f02e) involved simply passing the already captured `stacktrace` to [`Exception.normalize/3`](https://github.com/elixir-lang/elixir/blob/27f046321d8fe7e320cf8e0665d41d908f97a106/lib/elixir/lib/exception.ex#L97-L109).  I still wondered, however, where the `Keyword.delete_key/3` was coming from and decided to keep digging.

The [current implementation](https://github.com/elixir-lang/elixir/blob/e9dfa50c74488000d2c9de71e926cdd78609b3ac/lib/elixir/lib/keyword.ex#L467-L468) of `Keyword.delete/2` (and some other related functions) uses the throw/catch pattern to short-circuit for cases when the given key is not present within a list.  Since [`Keyword.put/3`](https://github.com/elixir-lang/elixir/blob/e9dfa50c74488000d2c9de71e926cdd78609b3ac/lib/elixir/lib/keyword.ex#L555) calls `Keyword.delete/2`, the stacktrace will be corrupted for lists where the key being "put" is not yet present.

I was hesitant to even submit this pull request due to the changes in syntax coming in OTP 21 related to catching errors and stacktraces.  And really, as a developer if the _first thing_ you do after a `catch` doesn't involve running `:erlang.get_stacktrace()` there's really no guarantee that some other call won't have corrupted the stacktrace by the time you get around to asking for it.

However, I do think this behavior is fairly confusing and is non-obvious where the stacktrace change is actually coming from.

So...

¯\\\_(ツ)\_/¯

#### A few options...

Therefore, I set out to undo all of the hard work @michalmuskala contributed with elixir-lang/elixir#6564 and possibly find a happy solution for everyone.

*TL;DR* This pull request currently contains an implementation described below as [`delete_v2/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L245-L260).  Alternatively, [`delete_v3/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L262-L280) has some pros/cons that might make it the preferred implementation.

I put together a very rough module to help measure the difference in time, reductions, and garbage collected heap bytes: https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e

Here's part of the output:

```
## count=100000, size=10
[h] delete_v0/2:     1.60 μs (avg),   344.05 gc/μs (avg),    41.04 reds/μs (avg)
[m] delete_v0/2:     1.82 μs (avg),   311.45 gc/μs (avg),    38.56 reds/μs (avg)
[h] delete_v1/2:     2.14 μs (avg),   232.09 gc/μs (avg),    19.50 reds/μs (avg)
[m] delete_v1/2:     1.19 μs (avg),   397.58 gc/μs (avg),    16.42 reds/μs (avg)
[h] delete_v2/2:     1.12 μs (avg),   569.17 gc/μs (avg),    17.59 reds/μs (avg)
[m] delete_v2/2:     1.44 μs (avg),   467.06 gc/μs (avg),    16.72 reds/μs (avg)
[h] delete_v3/2:     1.55 μs (avg),   413.40 gc/μs (avg),    16.92 reds/μs (avg)
[m] delete_v3/2:     1.02 μs (avg),   344.93 gc/μs (avg),     5.99 reds/μs (avg)
## count=100000, size=100
[h] delete_v0/2:     7.39 μs (avg),   269.51 gc/μs (avg),    71.53 reds/μs (avg)
[m] delete_v0/2:     7.11 μs (avg),   282.24 gc/μs (avg),    72.44 reds/μs (avg)
[h] delete_v1/2:     3.95 μs (avg),   489.96 gc/μs (avg),    96.15 reds/μs (avg)
[m] delete_v1/2:     2.41 μs (avg),   195.41 gc/μs (avg),    57.20 reds/μs (avg)
[h] delete_v2/2:     3.51 μs (avg),  1003.66 gc/μs (avg),    59.73 reds/μs (avg)
[m] delete_v2/2:     3.46 μs (avg),  1025.16 gc/μs (avg),    60.34 reds/μs (avg)
[h] delete_v3/2:     3.62 μs (avg),   971.51 gc/μs (avg),    55.39 reds/μs (avg)
[m] delete_v3/2:     1.02 μs (avg),   343.82 gc/μs (avg),     5.98 reds/μs (avg)
## count=100000, size=10000
[h] delete_v0/2:   312.93 μs (avg),   512.56 gc/μs (avg),   130.60 reds/μs (avg)
[m] delete_v0/2:   321.35 μs (avg),   499.19 gc/μs (avg),   128.00 reds/μs (avg)
[h] delete_v1/2:   138.01 μs (avg),  1161.84 gc/μs (avg),   149.91 reds/μs (avg)
[m] delete_v1/2:   150.13 μs (avg),     3.17 gc/μs (avg),    68.16 reds/μs (avg)
[h] delete_v2/2:   126.68 μs (avg),  2234.20 gc/μs (avg),    94.37 reds/μs (avg)
[m] delete_v2/2:   126.99 μs (avg),  2228.79 gc/μs (avg),    93.90 reds/μs (avg)
[h] delete_v3/2:   154.80 μs (avg),  1828.32 gc/μs (avg),    75.68 reds/μs (avg)
[m] delete_v3/2:    21.35 μs (avg),    16.67 gc/μs (avg),     0.29 reds/μs (avg)
```

| Symbol | Description |
|--------|-------------|
| `[h]`  | "hit" test where the very last element of a list was deleted |
| `[m]`  | "miss" test where no elements were deleted |
| `gc/μs` | the total(-ish) heap bytes garbage collected across all iterations divided by total time |
| `reds/μs` | the total reductions counted across all iterations divided by total time |

Here's a summary of the different `delete/2` implementations:

| Function | Description | Pros | Cons |
|----------|-------------|------|------|
| [`delete_v0/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L225-L228) | Uses [`lists:filter/2`](http://erlang.org/doc/man/lists.html#filter-2) to remove `key` (on `master` previously) | Memory Efficient | Slow for "hit" and "miss" |
| [`delete_v1/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L230-L243) | Uses throw/catch (on `master` now) | Faster than `delete_v0/2` for "hit" and "miss", More Memory Efficient for "miss" | Corrupts stacktrace, Slower than `delete_v2/2` and `delete_v3/2` for "miss" |
| [`delete_v2/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L245-L260) | Uses [`:lists.reverse/1,2`](http://erlang.org/doc/man/lists.html#reverse-1) and simple recursion (the version in this PR) | Faster than `delete_v1/2` for large lists | Less Memory Efficient |
| [`delete_v3/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L262-L280) | Same as `delete_v2/2` with short-circuit BIF [`:lists.keymember/3`](http://erlang.org/doc/man/lists.html#keymember-3) | Fastest on "miss", Moderately Memory Efficient | Slightly slower than `delete_v2/2` on "hit" for large lists |

Also, for the curious, I put together a [native C implementation](https://github.com/potatosalad/erlang-nativelists) to see if I could achieve better performance, but was unsuccessful:

```
## count=100000, size=10
[h] delete_na/2:     2.97 μs (avg),   166.02 gc/μs (avg),     2.85 reds/μs (avg)
[m] delete_na/2:     2.49 μs (avg),   205.61 gc/μs (avg),     3.08 reds/μs (avg)
## count=100000, size=100
[h] delete_na/2:     5.28 μs (avg),   358.11 gc/μs (avg),     2.70 reds/μs (avg)
[m] delete_na/2:     6.37 μs (avg),   278.72 gc/μs (avg),     2.52 reds/μs (avg)
## count=100000, size=10000
[h] delete_na/2:   172.08 μs (avg),   853.65 gc/μs (avg),     3.50 reds/μs (avg)
[m] delete_na/2:   143.41 μs (avg),   862.68 gc/μs (avg),     3.84 reds/μs (avg)
```

### CODE REVIEW

**Comment by michalmuskala** [comment]
We're going to be deprecating `System.get_stacktrace` in favour of `__STACKTRACE__` that's going to always do the right thing and grab the stack-trace first thing in the rescue/catch clause, so this can't happen.

**Comment by potatosalad** [comment]
@michalmuskala Awesome, I noticed some changes recently related to `__STACKTRACE__`, which was part of my hesitation in even submitting this pull request.

I am curious, however, if anyone thinks the [`delete_v3/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L262-L280) short-circuit method is beneficial enough on its own to merit inclusion.  For large lists, it would greatly speed up things like `Keyword.put/3` for lists where the key being "put" isn't present.

**Comment by michalmuskala** [comment]
I guess an interesting experiment would be to try running some trace over some example programs - compilation of some modules, a simple plug app, etc, to try to see if hit or miss is the common case for Keyword.

Additionally, I would generally optimise keyword for small lists rather than large - it's expected keyword lists shouldn't be large in principle.

**Comment by potatosalad** [comment]
> Additionally, I would generally optimise keyword for small lists rather than large - it's expected keyword lists shouldn't be large in principle.

I agree.  I probably should have phrased my statement more like this: "It's faster for both small and large lists when the key isn't present, but it's more noticeable on larger lists."

**Comment by potatosalad** [comment]
> I guess an interesting experiment would be to try running some trace over some example programs - compilation of some modules, a simple plug app, etc, to try to see if hit or miss is the common case for Keyword.

I put together a rough (very rough) trace on the 3 delete related functions touched by this pull request.  I ran the following with the trace-enabled compiled version of Elixir:

```bash
$ mix phx.new fake --no-ecto
$ cd fake
$ iex -S mix phx.server
```

The only function called was `&Keyword.delete/2`:

```
calls: 181
hit:    60
miss:  121
```

**Comment by michalmuskala** [comment]
There's another possible implementation. Have you considered it?

```elixir
  def delete_v4(keywords, key) when is_list(keywords) and is_atom(key) do
    case :lists.keymember(key, 1, keywords) do
      true -> delete_v4_key(keywords, key)
      _ -> keywords
    end
  end

  defp delete_v4_key([{key, _} | tail], key) do
    delete_v4_key(tail, key)
  end

  defp delete_v4_key([{_, _} = pair | tail], key) do
    [pair | delete_v4_key(tail, key)]
  end

  defp delete_v4_key([], _key) do
    []
  end
```

**Comment by potatosalad** [comment]
I just added [`delete_v4/2`](https://gist.github.com/potatosalad/f4fea8490bea09eb000a3a346a52cb6e#file-keyword_delete_bench-ex-L285-L303) to the gist and re-ran the benchmark:

```
## count=100000, size=10
[h] delete_v0/2:     1.56 μs (avg),   353.70 gc/μs (avg),    41.13 reds/μs (avg)
[m] delete_v0/2:     1.91 μs (avg),   296.56 gc/μs (avg),    38.38 reds/μs (avg)
[h] delete_v1/2:     2.13 μs (avg),   232.77 gc/μs (avg),    19.46 reds/μs (avg)
[m] delete_v1/2:     1.16 μs (avg),   406.83 gc/μs (avg),    16.46 reds/μs (avg)
[h] delete_v2/2:     1.10 μs (avg),   580.03 gc/μs (avg),    17.61 reds/μs (avg)
[m] delete_v2/2:     1.20 μs (avg),   560.04 gc/μs (avg),    17.38 reds/μs (avg)
[h] delete_v3/2:     1.10 μs (avg),   583.97 gc/μs (avg),    18.62 reds/μs (avg)
[m] delete_v3/2:     1.01 μs (avg),   349.55 gc/μs (avg),     5.99 reds/μs (avg)
[h] delete_v4/2:     2.15 μs (avg),   230.79 gc/μs (avg),    19.43 reds/μs (avg)
[m] delete_v4/2:     1.00 μs (avg),   350.56 gc/μs (avg),     5.99 reds/μs (avg)
## count=100000, size=100
[h] delete_v0/2:     6.80 μs (avg),   292.74 gc/μs (avg),    73.88 reds/μs (avg)
[m] delete_v0/2:     6.85 μs (avg),   293.00 gc/μs (avg),    72.89 reds/μs (avg)
[h] delete_v1/2:     3.77 μs (avg),   514.05 gc/μs (avg),    98.71 reds/μs (avg)
[m] delete_v1/2:     2.30 μs (avg),   205.54 gc/μs (avg),    59.24 reds/μs (avg)
[h] delete_v2/2:     3.32 μs (avg),  1061.19 gc/μs (avg),    61.16 reds/μs (avg)
[m] delete_v2/2:     3.31 μs (avg),  1073.99 gc/μs (avg),    60.90 reds/μs (avg)
[h] delete_v3/2:     3.53 μs (avg),   997.14 gc/μs (avg),    55.78 reds/μs (avg)
[m] delete_v3/2:     1.01 μs (avg),   348.60 gc/μs (avg),     5.98 reds/μs (avg)
[h] delete_v4/2:     3.80 μs (avg),   509.64 gc/μs (avg),    90.84 reds/μs (avg)
[m] delete_v4/2:     1.01 μs (avg),   349.54 gc/μs (avg),     5.99 reds/μs (avg)
## count=100000, size=10000
[h] delete_v0/2:   303.92 μs (avg),   527.75 gc/μs (avg),   134.67 reds/μs (avg)
[m] delete_v0/2:   305.38 μs (avg),   525.28 gc/μs (avg),   134.01 reds/μs (avg)
[h] delete_v1/2:   129.10 μs (avg),  1242.00 gc/μs (avg),   157.09 reds/μs (avg)
[m] delete_v1/2:   141.72 μs (avg),     3.36 gc/μs (avg),    71.43 reds/μs (avg)
[h] delete_v2/2:   122.49 μs (avg),  2310.56 gc/μs (avg),    96.92 reds/μs (avg)
[m] delete_v2/2:   122.97 μs (avg),  2301.67 gc/μs (avg),    96.52 reds/μs (avg)
[h] delete_v3/2:   150.64 μs (avg),  1878.88 gc/μs (avg),    77.73 reds/μs (avg)
[m] delete_v3/2:    20.90 μs (avg),    17.03 gc/μs (avg),     0.29 reds/μs (avg)
[h] delete_v4/2:   161.10 μs (avg),   995.29 gc/μs (avg),   125.84 reds/μs (avg)
[m] delete_v4/2:    20.76 μs (avg),    17.15 gc/μs (avg),     0.29 reds/μs (avg)
```

Slightly slower, but roughly twice as memory efficient.  Seems like a good improvement to me.

**Comment by michalmuskala** [comment]
I think we should go with the `v4` version - it's still significantly faster than the previous implementation and only slightly slower than the current implementation in the success case, but much faster in the miss case.

**Comment by michalmuskala** [comment]
Thank you! :heart:
